atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/BucketBasedTargetedSweeper.java

```diff
@@ -0,0 +1,262 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts;
+
+import com.google.common.collect.ImmutableList;
+import com.palantir.atlasdb.cleaner.PuncherStore;
+import com.palantir.atlasdb.sweep.BackgroundSweeper;
+import com.palantir.atlasdb.sweep.asts.bucketingthings.BucketWriter;
+import com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketAssigner;
+import com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketAssigner.BucketAssignerEventHandler;
+import com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketCloseTimestampCalculator;
+import com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultBucketWriter;
+import com.palantir.atlasdb.sweep.asts.bucketingthings.DefaultSweepAssignedBucketStore;
+import com.palantir.atlasdb.sweep.asts.locks.LockableFactory;
+import com.palantir.atlasdb.sweep.asts.progress.BucketProgressStore;
+import com.palantir.atlasdb.sweep.metrics.TargetedSweepMetrics;
+import com.palantir.atlasdb.sweep.queue.ShardAndStrategy;
+import com.palantir.atlasdb.sweep.queue.SpecialTimestampsSupplier;
+import com.palantir.atlasdb.sweep.queue.SweepQueueCleaner;
+import com.palantir.atlasdb.sweep.queue.SweepQueueComponents;
+import com.palantir.atlasdb.sweep.queue.SweepQueueProgressUpdater;
+import com.palantir.atlasdb.sweep.queue.config.AutoScalingTargetedSweepRuntimeConfig;
+import com.palantir.atlasdb.table.description.SweeperStrategy;
+import com.palantir.common.concurrent.PTExecutors;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.refreshable.Refreshable;
+import java.time.Duration;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public final class BucketBasedTargetedSweeper implements BackgroundSweeper {
+    private final ScheduledExecutorService backgroundTaskExecutor;
+    // Separating the executors so that lag on the background executor does not affect the sweepers.
+    private final ScheduledExecutorService sweeperTaskExecutor;
+    private final List<DynamicTaskScheduler> scheduledTasks;
+
+    private BucketBasedTargetedSweeper(
+            ScheduledExecutorService backgroundTaskExecutor,
+            ScheduledExecutorService sweeperTaskExecutor,
+            List<DynamicTaskScheduler> scheduledTasks) {
+        this.backgroundTaskExecutor = backgroundTaskExecutor;
+        this.sweeperTaskExecutor = sweeperTaskExecutor;
+        this.scheduledTasks = List.copyOf(scheduledTasks);
+    }
+
+    // TODO(mdaudali): Will address this later. It's possible I just inject the created components in.
+    //  but I think this will change with some of the thread autoscaling work so deferring.
```

Comment:
non-actionable: Is fine. I would normally have a factory or factories given the complexity of this construction, but fine to hold off for now